File: 113. Path Sum II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        DFS(res, {}, root, sum);
        return res;
    }
    
    void DFS(vector<vector<int>>& res, vector <int> path, TreeNode* root, int sum)
    {
        if (root == NULL)
            return;
        if (root->left == NULL && root->right == NULL && sum == root->val)
        {
            path.push_back(root->val);
            res.push_back(path);
            path.pop_back();
            return;
        }
        path.push_back(root->val);
        DFS(res, path, root->left, sum - root->val);
        DFS(res, path, root->right, sum - root->val);
        path.pop_back();
    }
};

File: 113. Path Sum II.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function(root, sum) {
    var res = [];
    var DFS = function(path, root, sum)
    {
        if (root === null)
            return;
        if (root.left === null && root.right === null && sum == root.val)
        {
            path.push(root.val);
            res.push([]);
            for (var i = 0; i < path.length; i++)
                res[res.length-1].push(path[i]);
            path.pop();
            return;
        }
        path.push(root.val);
        DFS(path, root.left, sum - root.val);
        DFS(path, root.right, sum - root.val);
        path.pop();
    };
    DFS([], root, sum);
    return res;
};
